server/src/main/java/org/opensearch/action/pagination/ShardPaginationStrategy.java

```diff
@@ -0,0 +1,273 @@
+                PageParams.PARAM_ASC_SORT_VALUE.equals(sortOrder) ? ASC_COMPARATOR : DESC_COMPARATOR
+            );
+        } else {
+            return PaginationStrategy.getSortedIndexMetadata(
+                clusterState,
+                getMetadataFilter(sortOrder, lastIndexName, lastIndexCreationTime),
+                PageParams.PARAM_ASC_SORT_VALUE.equals(sortOrder) ? ASC_COMPARATOR : DESC_COMPARATOR
+            );
+        }
+    }
+
+    private static Predicate<IndexMetadata> getMetadataFilter(String sortOrder, String lastIndexName, Long lastIndexCreationTime) {
+        if (Objects.isNull(lastIndexName) || Objects.isNull(lastIndexCreationTime)) {
+            return indexMetadata -> true;
+        }
+        return indexNameFilter(lastIndexName).or(
+            IndexPaginationStrategy.getIndexCreateTimeFilter(sortOrder, lastIndexName, lastIndexCreationTime)
+        );
+    }
+
+    private static Predicate<IndexMetadata> indexNameFilter(String lastIndexName) {
+        return metadata -> metadata.getIndex().getName().equals(lastIndexName);
+    }
+
+    /**
+     * Will be used to get the list of shards and respective indices to which they belong,
+     * which are to be displayed in a page.
+     * Note: All shards for a shardID will always be present in the same page.
+     */
+    private PageData getPageData(
+        List<IndexMetadata> filteredIndices,
+        Map<String, IndexRoutingTable> indicesRouting,
+        final ShardStrategyToken token,
+        final int numShardsRequired
+    ) {
+        List<ShardRouting> shardRoutings = new ArrayList<>();
+        List<String> indices = new ArrayList<>();
+        int shardCount = 0;
+        IndexMetadata lastAddedIndex = null;
+
+        // iterate over indices until shardCount is less than numShardsRequired
+        for (IndexMetadata indexMetadata : filteredIndices) {
+            String indexName = indexMetadata.getIndex().getName();
+            boolean indexShardsAdded = false;
+            // Always start from shardID 0 for all indices except for the first one which might be same as the last sent
+            // index. To identify if an index is same as last sent index, verify both the index name and creaton time.
+            int startShardId = shardCount == 0 ? getStartShardIdForPageIndex(token, indexName, indexMetadata.getCreationDate()) : 0;
+            Map<Integer, IndexShardRoutingTable> indexShardRoutingTable = indicesRouting.get(indexName).getShards();
+            for (; startShardId < indexShardRoutingTable.size(); startShardId++) {
+                if (indexShardRoutingTable.get(startShardId).size() > numShardsRequired) {
+                    throw new IllegalArgumentException("size value should be greater than the replica count of all indices");
+                }
+                shardCount += indexShardRoutingTable.get(startShardId).size();
+                if (shardCount > numShardsRequired) {
+                    break;
```

Comment:
you are breaking from for loop by incrementing the shardCount but ShardRoutings were not added to ```shardRoutings``` object. this is to ensure we add shards <= size limit from request?

Reply:
This is done to ensure that shardRoutings for a shardID do not get split across pages. Basically, if running ```shardCount``` is exceeding the ```pageSize/numShardsRequired``` because of a shardID then that shardID should be included in the next page.